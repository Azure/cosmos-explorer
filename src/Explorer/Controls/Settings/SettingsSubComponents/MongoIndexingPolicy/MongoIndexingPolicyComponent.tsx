import * as React from "react";
import {
  DetailsList,
  DetailsListLayoutMode,
  Stack,
  IconButton,
  Text,
  SelectionMode,
  IColumn,
  MessageBar,
  MessageBarType,
  Spinner,
  SpinnerSize,
  Separator,
} from "@fluentui/react";
import {
  addMongoIndexStackProps,
  customDetailsListStyles,
  mongoIndexingPolicyDisclaimer,
  mediumWidthStackStyles,
  subComponentStackProps,
  createAndAddMongoIndexStackProps,
  separatorStyles,
  indexingPolicynUnsavedWarningMessage,
  infoAndToolTipTextStyle,
  onRenderRow,
  mongoCompoundIndexNotSupportedMessage,
} from "../../SettingsRenderUtils";
import { MongoIndex } from "../../../../../Utils/arm/generatedClients/2020-04-01/types";
import {
  MongoIndexTypes,
  AddMongoIndexProps,
  MongoIndexIdField,
  MongoNotificationType,
  getMongoIndexType,
  getMongoIndexTypeText,
  isIndexTransforming,
} from "../../SettingsUtils";
import { AddMongoIndexComponent } from "./AddMongoIndexComponent";
import { CollapsibleSectionComponent } from "../../../CollapsiblePanel/CollapsibleSectionComponent";
import { IndexingPolicyRefreshComponent } from "../IndexingPolicyRefresh/IndexingPolicyRefreshComponent";

export interface MongoIndexingPolicyComponentProps {
  mongoIndexes: MongoIndex[];
  onIndexDrop: (index: number) => void;
  indexesToDrop: number[];
  onRevertIndexDrop: (index: number) => void;
  indexesToAdd: AddMongoIndexProps[];
  onRevertIndexAdd: (index: number) => void;
  onIndexAddOrChange: (index: number, description: string, type: MongoIndexTypes) => void;
  indexTransformationProgress: number;
  refreshIndexTransformationProgress: () => Promise<void>;
  onMongoIndexingPolicySaveableChange: (isMongoIndexingPolicySaveable: boolean) => void;
  onMongoIndexingPolicyDiscardableChange: (isMongoIndexingPolicyDiscardable: boolean) => void;
}

interface MongoIndexDisplayProps {
  definition: JSX.Element;
  type: JSX.Element;
  actionButton: JSX.Element;
}

export class MongoIndexingPolicyComponent extends React.Component<MongoIndexingPolicyComponentProps> {
  private shouldCheckComponentIsDirty = true;
  private addMongoIndexComponentRefs: React.RefObject<AddMongoIndexComponent>[] = [];
  private initialIndexesColumns: IColumn[] = [
    { key: "definition", name: "Definition", fieldName: "definition", minWidth: 100, maxWidth: 200, isResizable: true },
    { key: "type", name: "Type", fieldName: "type", minWidth: 100, maxWidth: 200, isResizable: true },
    {
      key: "actionButton",
      name: "Drop Index",
      fieldName: "actionButton",
      minWidth: 100,
      maxWidth: 200,
      isResizable: true,
    },
  ];

  private indexesToBeDroppedColumns: IColumn[] = [
    { key: "definition", name: "Definition", fieldName: "definition", minWidth: 100, maxWidth: 200, isResizable: true },
    { key: "type", name: "Type", fieldName: "type", minWidth: 100, maxWidth: 200, isResizable: true },
    {
      key: "actionButton",
      name: "Add index back",
      fieldName: "actionButton",
      minWidth: 100,
      maxWidth: 200,
      isResizable: true,
    },
  ];

  componentDidUpdate(prevProps: MongoIndexingPolicyComponentProps): void {
    if (this.props.indexesToAdd.length > prevProps.indexesToAdd.length) {
      this.addMongoIndexComponentRefs[prevProps.indexesToAdd.length]?.current?.focus();
    }
    this.onComponentUpdate();
  }

  componentDidMount(): void {
    this.onComponentUpdate();
  }

  private onComponentUpdate = (): void => {
    if (!this.shouldCheckComponentIsDirty) {
      this.shouldCheckComponentIsDirty = true;
      return;
    }
    this.props.onMongoIndexingPolicySaveableChange(this.isMongoIndexingPolicySaveable());
    this.props.onMongoIndexingPolicyDiscardableChange(this.isMongoIndexingPolicyDiscardable());
    this.shouldCheckComponentIsDirty = false;
  };

  public isMongoIndexingPolicySaveable = (): boolean => {
    if (this.props.indexesToAdd.length === 0 && this.props.indexesToDrop.length === 0) {
      return false;
    }

    const addErrorsExist = !!this.props.indexesToAdd.find((addMongoIndexProps) => addMongoIndexProps.notification);

    if (addErrorsExist) {
      return false;
    }

    return true;
  };

  public isMongoIndexingPolicyDiscardable = (): boolean => {
    return this.props.indexesToAdd.length > 0 || this.props.indexesToDrop.length > 0;
  };

  public getMongoWarningNotificationMessage = (): JSX.Element => {
    const warningMessage = this.props.indexesToAdd.find(
      (addMongoIndexProps) => addMongoIndexProps.notification?.type === MongoNotificationType.Warning
    )?.notification.message;

    if (warningMessage) {
      return <Text styles={infoAndToolTipTextStyle}>{warningMessage}</Text>;
    }
    return undefined;
  };

  private getActionButton = (arrayPosition: number, isCurrentIndex: boolean): JSX.Element => {
    return isCurrentIndex ? (
      <IconButton
        ariaLabel="Delete index Button"
        iconProps={{ iconName: "Delete" }}
        disabled={isIndexTransforming(this.props.indexTransformationProgress)}
        onClick={() => {
          this.props.onIndexDrop(arrayPosition);
        }}
      />
    ) : (
      <IconButton
        ariaLabel="Add back Index Button"
        iconProps={{ iconName: "Add" }}
        onClick={() => {
          this.props.onRevertIndexDrop(arrayPosition);
        }}
      />
    );
  };

  private getMongoIndexDisplayProps = (
    mongoIndex: MongoIndex,
    arrayPosition: number,
    isCurrentIndex: boolean
  ): MongoIndexDisplayProps => {
    const keys = mongoIndex?.key?.keys;
    const type = getMongoIndexType(keys);
    const definition = keys?.join();
    let mongoIndexDisplayProps: MongoIndexDisplayProps;
    if (type) {
      mongoIndexDisplayProps = {
        definition: <Text>{definition}</Text>,
        type: <Text>{getMongoIndexTypeText(type)}</Text>,
        actionButton: definition === MongoIndexIdField ? <></> : this.getActionButton(arrayPosition, isCurrentIndex),
      };
    }
    return mongoIndexDisplayProps;
  };

  private renderIndexesToBeAdded = (): JSX.Element => {
    const indexesToAddLength = this.props.indexesToAdd.length;
    for (let i = 0; i < indexesToAddLength; i++) {
      const existingIndexToAddRef = React.createRef<AddMongoIndexComponent>();
      this.addMongoIndexComponentRefs[i] = existingIndexToAddRef;
    }
    const newIndexToAddRef = React.createRef<AddMongoIndexComponent>();
    this.addMongoIndexComponentRefs[indexesToAddLength] = newIndexToAddRef;

    return (
      <Stack {...addMongoIndexStackProps} styles={mediumWidthStackStyles}>
        {this.props.indexesToAdd.map((mongoIndexWithType, arrayPosition) => {
          const keys = mongoIndexWithType.mongoIndex.key.keys;
          const type = mongoIndexWithType.type;
          const notification = mongoIndexWithType.notification;
          return (
            <AddMongoIndexComponent
              ref={this.addMongoIndexComponentRefs[arrayPosition]}
              position={arrayPosition}
              key={arrayPosition}
              description={keys.join()}
              type={type}
              notification={notification}
              onIndexAddOrChange={(description, type) =>
                this.props.onIndexAddOrChange(arrayPosition, description, type)
              }
              onDiscard={() => {
                this.addMongoIndexComponentRefs.splice(arrayPosition, 1);
                this.props.onRevertIndexAdd(arrayPosition);
              }}
            />
          );
        })}

        <AddMongoIndexComponent
          ref={this.addMongoIndexComponentRefs[indexesToAddLength]}
          disabled={isIndexTransforming(this.props.indexTransformationProgress)}
          position={indexesToAddLength}
          key={indexesToAddLength}
          description={undefined}
          type={undefined}
          notification={undefined}
          onIndexAddOrChange={(description, type) =>
            this.props.onIndexAddOrChange(indexesToAddLength, description, type)
          }
          onDiscard={() => {
            this.props.onRevertIndexAdd(indexesToAddLength);
          }}
        />
      </Stack>
    );
  };

  private renderInitialIndexes = (): JSX.Element => {
    const initialIndexes = this.props.mongoIndexes
      .map((mongoIndex, arrayPosition) => this.getMongoIndexDisplayProps(mongoIndex, arrayPosition, true))
      .filter((value, arrayPosition) => !!value && !this.props.indexesToDrop.includes(arrayPosition));

    return (
      <Stack {...createAndAddMongoIndexStackProps} styles={mediumWidthStackStyles}>
        <CollapsibleSectionComponent title="Current index(es)" isExpandedByDefault={true}>
          {
            <>
              <DetailsList
                styles={customDetailsListStyles}
                disableSelectionZone
                items={initialIndexes}
                columns={this.initialIndexesColumns}
                selectionMode={SelectionMode.none}
                onRenderRow={onRenderRow}
                layoutMode={DetailsListLayoutMode.justified}
              />
              {this.renderIndexesToBeAdded()}
            </>
          }
        </CollapsibleSectionComponent>
      </Stack>
    );
  };

  private renderIndexesToBeDropped = (): JSX.Element => {
    const indexesToBeDropped = this.props.indexesToDrop.map((dropIndex, arrayPosition) =>
      this.getMongoIndexDisplayProps(this.props.mongoIndexes[dropIndex], arrayPosition, false)
    );

    return (
      <Stack styles={mediumWidthStackStyles}>
        <CollapsibleSectionComponent title="Index(es) to be dropped" isExpandedByDefault={true}>
          {indexesToBeDropped.length > 0 && (
            <DetailsList
              styles={customDetailsListStyles}
              disableSelectionZone
              items={indexesToBeDropped}
              columns={this.indexesToBeDroppedColumns}
              selectionMode={SelectionMode.none}
              onRenderRow={onRenderRow}
              layoutMode={DetailsListLayoutMode.justified}
            />
          )}
        </CollapsibleSectionComponent>
      </Stack>
    );
  };

  public hasCompoundIndex = (): boolean => {
    for (let index = 0; index < this.props.mongoIndexes.length; index++) {
      if (this.props.mongoIndexes[index].key?.keys?.length > 1) {
        return true;
      }
    }
    return false;
  };

  private renderWarningMessage = (): JSX.Element => {
    let warningMessage: JSX.Element;
    if (this.getMongoWarningNotificationMessage()) {
      warningMessage = this.getMongoWarningNotificationMessage();
    } else if (this.isMongoIndexingPolicySaveable()) {
      warningMessage = indexingPolicynUnsavedWarningMessage;
    }

    return (
      <>
        <IndexingPolicyRefreshComponent
          indexTransformationProgress={this.props.indexTransformationProgress}
          refreshIndexTransformationProgress={this.props.refreshIndexTransformationProgress}
        />
        {warningMessage && <MessageBar messageBarType={MessageBarType.warning}>{warningMessage}</MessageBar>}
      </>
    );
  };

  public render(): JSX.Element {
    if (this.props.mongoIndexes) {
      if (this.hasCompoundIndex()) {
        return mongoCompoundIndexNotSupportedMessage;
      }
      return (
        <Stack {...subComponentStackProps}>
          {this.renderWarningMessage()}
          {mongoIndexingPolicyDisclaimer}
          {this.renderInitialIndexes()}
          <Separator styles={separatorStyles} />
          {this.renderIndexesToBeDropped()}
        </Stack>
      );
    } else {
      return <Spinner size={SpinnerSize.large} />;
    }
  }
}
