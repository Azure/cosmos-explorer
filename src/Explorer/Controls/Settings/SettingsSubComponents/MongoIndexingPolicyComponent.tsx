import * as React from "react";
import {
  DetailsList,
  DetailsListLayoutMode,
  MessageBar,
  MessageBarType,
  Stack,
  IconButton,
  Text,
  SelectionMode,
  IDetailsRowProps,
  IDetailsRowStyles,
  DetailsRow,
  Label,
  TextField,
  IDropdownOption,
  Dropdown
} from "office-ui-fabric-react";
import { titleAndInputStackProps } from "../SettingsRenderUtils";
import { MongoIndex } from "../../../../Utils/arm/generatedClients/2020-04-01/types";
import { MongoIndexTypes } from "../SettingsUtils";

export interface MongoIndexingPolicyComponentProps {
  mongoIndexes: MongoIndex[];
  onIndexDelete: (index: number) => void;
  indexesToDelete: number[];
  onRevertIndexDelete: (index: number) => void;
  indexesToAdd: MongoIndex[];
  onRevertIndexAdd: (index: number) => void;
  onIndexAdd: (newMongoIndex: MongoIndex) => void;
  onMongoIndexingPolicyDirtyChange: (isMongoIndexingPolicyDirty: boolean) => void;
}

interface MongoIndexingPolicyComponentState {
  isAddingIndex: boolean;
  newIndexDefinition: string;
  newIndexType: MongoIndexTypes;
  errorMessage: string;
}

interface MongoIndexDisplayProps {
  definition: JSX.Element;
  type: JSX.Element;
  dropIndex: JSX.Element;
}
export class MongoIndexingPolicyComponent extends React.Component<
  MongoIndexingPolicyComponentProps,
  MongoIndexingPolicyComponentState
> {
  private shouldCheckComponentIsDirty = true;
  private columns = [
    { key: "definition", name: "Definition", fieldName: "definition", minWidth: 100, maxWidth: 200, isResizable: true },
    { key: "type", name: "Type", fieldName: "type", minWidth: 100, maxWidth: 200, isResizable: true },
    { key: "dropIndex", name: "Drop Index", fieldName: "dropIndex", minWidth: 100, maxWidth: 200, isResizable: true }
  ];
  private indexTypes: IDropdownOption[] = [MongoIndexTypes.Single, MongoIndexTypes.WildCard].map((value: string) => ({
    text: value,
    key: value
  }));

  constructor(props: MongoIndexingPolicyComponentProps) {
    super(props);
    this.state = {
      isAddingIndex: false,
      newIndexDefinition: undefined,
      newIndexType: undefined,
      errorMessage: undefined
    };
  }

  componentDidUpdate(): void {
    this.onComponentUpdate();
  }

  componentDidMount(): void {
    this.onComponentUpdate();
  }

  private onComponentUpdate = (): void => {
    if (!this.shouldCheckComponentIsDirty) {
      this.shouldCheckComponentIsDirty = true;
      return;
    }
    this.props.onMongoIndexingPolicyDirtyChange(this.IsComponentDirty());
    this.shouldCheckComponentIsDirty = false;
  };

  public IsComponentDirty = (): boolean => {
    return this.props.indexesToAdd.length > 0 || this.props.indexesToDelete.length > 0;
  };

  private onRenderRow(props: IDetailsRowProps): JSX.Element {
    const rowStyles: Partial<IDetailsRowStyles> = {
      root: {
        selectors: {
          ":hover": {
            background: "transparent"
          }
        }
      }
    };

    return <DetailsRow {...props} styles={rowStyles} />;
  }

  public getType = (keys: string[]): string => {
    const length = keys.length;
    let type: string = undefined;
    if (length === 1) {
      if (keys[0].indexOf("$**") !== -1) {
        type = MongoIndexTypes.WildCard;
      } else {
        type = MongoIndexTypes.Single;
      }
    }
    return type;
  };

  public renderInitialIndexes = (): JSX.Element => {
    const initialIndexes: MongoIndexDisplayProps[] = this.props.mongoIndexes.map(
      (mongoIndex: MongoIndex, index: number) => {
        const keys = mongoIndex.key.keys;
        const type = this.getType(keys);
        const definition = keys.join();
        return type
          ? ({
              definition: <Text>{definition}</Text>,
              type: <Text>{type}</Text>,
              dropIndex:
                definition === "_id" ? (
                  <></>
                ) : (
                  <IconButton
                    disabled={this.props.indexesToDelete.includes(index)}
                    iconProps={{ iconName: "Delete" }}
                    onClick={() => {
                      this.props.onIndexDelete(index);
                    }}
                  />
                )
            } as MongoIndexDisplayProps)
          : undefined;
      }
    );

    return (
      <Stack {...titleAndInputStackProps}>
        <Label>Existing Indexes</Label>
        <DetailsList
          disableSelectionZone
          items={initialIndexes}
          columns={this.columns}
          selectionMode={SelectionMode.none}
          onRenderRow={this.onRenderRow}
          layoutMode={DetailsListLayoutMode.justified}
        />
      </Stack>
    );
  };

  public renderIndexesToBeDeleted = (): JSX.Element => {
    const indexesToBeDeleted: MongoIndexDisplayProps[] = this.props.indexesToDelete.map(
      (deleteIndex: number, index: number) => {
        const keys = this.props.mongoIndexes[deleteIndex].key.keys;
        const type = this.getType(keys);
        return {
          definition: <Text>{keys.join(",")}</Text>,
          type: <Text>{type}</Text>,
          dropIndex: (
            <IconButton
              iconProps={{ iconName: "Delete" }}
              onClick={() => {
                this.props.onRevertIndexDelete(index);
              }}
            />
          )
        } as MongoIndexDisplayProps;
      }
    );

    return indexesToBeDeleted.length != 0 ? (
      <Stack {...titleAndInputStackProps}>
        <Label>Indexes to be deleted</Label>
        <DetailsList
          disableSelectionZone
          items={indexesToBeDeleted}
          columns={this.columns}
          selectionMode={SelectionMode.none}
          onRenderRow={this.onRenderRow}
          layoutMode={DetailsListLayoutMode.justified}
        />
      </Stack>
    ) : (
      <></>
    );
  };

  public renderIndexesToBeAdded = (): JSX.Element => {
    const indexesToBeAdded: MongoIndexDisplayProps[] = this.props.indexesToAdd.map(
      (mongoIndex: MongoIndex, index: number) => {
        const keys = mongoIndex.key.keys;
        const type = this.getType(keys);
        return {
          definition: <Text>{keys.join(",")}</Text>,
          type: <Text>{type}</Text>,
          dropIndex: (
            <IconButton
              iconProps={{ iconName: "Delete" }}
              onClick={() => {
                this.props.onRevertIndexAdd(index);
              }}
            />
          )
        } as MongoIndexDisplayProps;
      }
    );

    return indexesToBeAdded.length != 0 ? (
      <Stack {...titleAndInputStackProps}>
        <Label>Indexes to be added</Label>
        <DetailsList
          disableSelectionZone
          items={indexesToBeAdded}
          columns={this.columns}
          selectionMode={SelectionMode.none}
          onRenderRow={this.onRenderRow}
          layoutMode={DetailsListLayoutMode.justified}
        />
      </Stack>
    ) : (
      <></>
    );
  };

  private onNewIndexDefinitionChange = (
    event: React.FormEvent<HTMLInputElement | HTMLTextAreaElement>,
    newValue?: string
  ): void => this.setState({ newIndexDefinition: newValue });

  private onNewIndexTypeChange = (
    event: React.FormEvent<HTMLDivElement>,
    option?: IDropdownOption,
    index?: number
  ): void => this.setState({ newIndexType: MongoIndexTypes[option.key as keyof typeof MongoIndexTypes] });

  public renderAddIndexComponent = (): JSX.Element => {
    return (
      <Stack {...titleAndInputStackProps}>
        <IconButton
          iconProps={{ iconName: "Add" }}
          disabled={this.state.isAddingIndex}
          onClick={() => {
            this.setState({ isAddingIndex: true });
          }}
        />
        {this.state.isAddingIndex && (
          <Stack horizontal tokens={{ childrenGap: 20 }}>
            <TextField
              styles={{ root: { width: 300 } }}
              label="Defintion"
              placeholder={this.state.newIndexType === MongoIndexTypes.WildCard ? "placeholder.$**" : undefined}
              onChange={this.onNewIndexDefinitionChange}
            />

            <Dropdown
              styles={{ dropdown: { width: 300 } }}
              placeholder="Select an index type"
              label="Type"
              options={this.indexTypes}
              onChange={this.onNewIndexTypeChange}
            />

            <IconButton
              iconProps={{ iconName: "Save" }}
              onClick={() => {
                if (this.state.newIndexType === undefined) {
                  this.setState({ errorMessage: "Please select the type of the index" });
                  return;
                }

                if (
                  this.state.newIndexType === MongoIndexTypes.WildCard &&
                  this.state.newIndexDefinition.indexOf("$**") === -1
                ) {
                  this.setState({ errorMessage: "Wild Card path is not present" });
                  return;
                }

                const newMongoIndex = { key: { keys: [this.state.newIndexDefinition] } } as MongoIndex;
                this.props.onIndexAdd(newMongoIndex);
                this.setState({
                  isAddingIndex: false,
                  newIndexDefinition: undefined,
                  newIndexType: undefined,
                  errorMessage: undefined
                });
              }}
            />
            <IconButton
              iconProps={{ iconName: "Cancel" }}
              onClick={() => {
                this.setState({
                  isAddingIndex: false,
                  newIndexDefinition: undefined,
                  newIndexType: undefined,
                  errorMessage: undefined
                });
              }}
            />
          </Stack>
        )}
        {this.state.errorMessage && (
          <MessageBar messageBarType={MessageBarType.error}>{this.state.errorMessage}</MessageBar>
        )}
      </Stack>
    );
  };

  public render(): JSX.Element {
    return (
      <Stack {...titleAndInputStackProps}>
        {this.renderInitialIndexes()}
        {this.renderIndexesToBeAdded()}
        {this.renderAddIndexComponent()}
        {this.renderIndexesToBeDeleted()}
      </Stack>
    );
  }
}
