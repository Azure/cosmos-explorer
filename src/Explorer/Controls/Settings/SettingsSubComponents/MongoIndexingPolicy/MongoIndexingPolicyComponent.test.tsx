import { shallow } from "enzyme";
import React from "react";
import { renderToString } from "react-dom/server";
import { MongoIndexTypes, MongoNotificationMessage, MongoNotificationType } from "../../SettingsUtils";
import { MongoIndexingPolicyComponent, MongoIndexingPolicyComponentProps } from "./MongoIndexingPolicyComponent";

describe("MongoIndexingPolicyComponent", () => {
  const baseProps: MongoIndexingPolicyComponentProps = {
    mongoIndexes: [],
    onIndexDrop: () => {
      return;
    },
    indexesToDrop: [],
    onRevertIndexDrop: () => {
      return;
    },
    indexesToAdd: [],
    onRevertIndexAdd: () => {
      return;
    },
    onIndexAddOrChange: () => {
      return;
    },
    indexTransformationProgress: undefined,
    refreshIndexTransformationProgress: () => new Promise(jest.fn()),
    onMongoIndexingPolicySaveableChange: () => {
      return;
    },
    onMongoIndexingPolicyDiscardableChange: () => {
      return;
    },
  };

  it("renders", () => {
    const wrapper = shallow(<MongoIndexingPolicyComponent {...baseProps} />);
    expect(wrapper).toMatchSnapshot();
  });

  it("error shown for collection with compound indexes", () => {
    const props = { ...baseProps, mongoIndexes: [{ key: { keys: ["prop1", "prop2"] } }] };
    const wrapper = shallow(<MongoIndexingPolicyComponent {...props} />);
    expect(wrapper).toMatchSnapshot();
    const mongoIndexingPolicyComponent = wrapper.instance() as MongoIndexingPolicyComponent;
    expect(mongoIndexingPolicyComponent.hasCompoundIndex()).toBeTruthy();
  });

  describe("AddMongoIndexProps test", () => {
    const wrapper = shallow(<MongoIndexingPolicyComponent {...baseProps} />);
    const mongoIndexingPolicyComponent = wrapper.instance() as MongoIndexingPolicyComponent;

    it("defaults", () => {
      expect(mongoIndexingPolicyComponent.isMongoIndexingPolicySaveable()).toEqual(false);
      expect(mongoIndexingPolicyComponent.isMongoIndexingPolicyDiscardable()).toEqual(false);
      expect(mongoIndexingPolicyComponent.getMongoWarningNotificationMessage()).toBeUndefined();
    });

    const sampleWarning = "sampleWarning";
    const sampleError = "sampleError";

    const cases = [
      [
        { type: MongoNotificationType.Warning, message: sampleWarning } as MongoNotificationMessage,
        false,
        false,
        true,
        sampleWarning,
      ],
      [
        { type: MongoNotificationType.Error, message: sampleError } as MongoNotificationMessage,
        false,
        false,
        true,
        undefined,
      ],
      [
        { type: MongoNotificationType.Error, message: sampleError } as MongoNotificationMessage,
        true,
        false,
        true,
        undefined,
      ],
      [undefined, false, true, true, undefined],
      [undefined, true, true, true, undefined],
    ];

    test.each(cases)(
      "mongo indexing policy saveable and discardable",
      (
        notification: MongoNotificationMessage,
        indexToDropIsPresent: boolean,
        isMongoIndexingPolicySaveable: boolean,
        isMongoIndexingPolicyDiscardable: boolean,
        mongoWarningNotificationMessage: string,
      ) => {
        const addMongoIndexProps = {
          mongoIndex: { key: { keys: ["sampleKey"] } },
          type: MongoIndexTypes.Single,
          notification: notification,
        };

        let indexesToDrop: number[] = [];
        if (indexToDropIsPresent) {
          indexesToDrop = [0];
        }
        wrapper.setProps({ indexesToAdd: [addMongoIndexProps], indexesToDrop: indexesToDrop });
        wrapper.update();

        expect(mongoIndexingPolicyComponent.isMongoIndexingPolicySaveable()).toEqual(isMongoIndexingPolicySaveable);
        expect(mongoIndexingPolicyComponent.isMongoIndexingPolicyDiscardable()).toEqual(
          isMongoIndexingPolicyDiscardable,
        );
        if (mongoWarningNotificationMessage) {
          const elementAsString = renderToString(mongoIndexingPolicyComponent.getMongoWarningNotificationMessage());
          // eslint-disable-next-line jest/no-conditional-expect
          expect(elementAsString).toContain(mongoWarningNotificationMessage);
        } else {
          // eslint-disable-next-line jest/no-conditional-expect
          expect(mongoIndexingPolicyComponent.getMongoWarningNotificationMessage()).toBeUndefined();
        }
      },
    );
  });
});
