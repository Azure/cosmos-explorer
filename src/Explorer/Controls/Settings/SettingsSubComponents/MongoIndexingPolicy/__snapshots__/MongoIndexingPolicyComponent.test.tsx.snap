// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MongoIndexingPolicyComponent error shown for collection with compound indexes 1`] = `
<Text>
  Collections with compound indexes are not yet supported in the indexing editor. To modify indexing policy for this collection, use the Mongo Shell.
</Text>
`;

exports[`MongoIndexingPolicyComponent renders 1`] = `
<Stack
  tokens={
    Object {
      "childrenGap": 20,
    }
  }
>
  <IndexingPolicyRefreshComponent
    refreshIndexTransformationProgress={[Function]}
  />
  <Text>
    For queries that filter on multiple properties, create multiple single field indexes instead of a compound index.
    <StyledLinkBase
      href="https://docs.microsoft.com/azure/cosmos-db/mongodb-indexing#index-types"
      target="_blank"
    >
       Compound indexes 
    </StyledLinkBase>
    are only used for sorting query results. If you need to add a compound index, you can create one using the Mongo shell.
  </Text>
  <Stack
    styles={
      Object {
        "root": Object {
          "width": 600,
        },
      }
    }
    tokens={
      Object {
        "childrenGap": 5,
      }
    }
  >
    <CollapsibleSectionComponent
      isExpandedByDefault={true}
      title="Current index(es)"
    >
      <StyledWithViewportComponent
        columns={
          Array [
            Object {
              "fieldName": "definition",
              "isResizable": true,
              "key": "definition",
              "maxWidth": 200,
              "minWidth": 100,
              "name": "Definition",
            },
            Object {
              "fieldName": "type",
              "isResizable": true,
              "key": "type",
              "maxWidth": 200,
              "minWidth": 100,
              "name": "Type",
            },
            Object {
              "fieldName": "actionButton",
              "isResizable": true,
              "key": "actionButton",
              "maxWidth": 200,
              "minWidth": 100,
              "name": "Drop Index",
            },
          ]
        }
        disableSelectionZone={true}
        items={Array []}
        layoutMode={1}
        onRenderRow={[Function]}
        selectionMode={0}
        styles={
          Object {
            "root": Object {
              "selectors": Object {
                ".ms-FocusZone": Object {
                  "paddingTop": 0,
                },
              },
            },
          }
        }
      />
      <Stack
        styles={
          Object {
            "root": Object {
              "width": 600,
            },
          }
        }
        tokens={
          Object {
            "childrenGap": 10,
          }
        }
      >
        <AddMongoIndexComponent
          disabled={false}
          key="0"
          onDiscard={[Function]}
          onIndexAddOrChange={[Function]}
          position={0}
        />
      </Stack>
    </CollapsibleSectionComponent>
  </Stack>
  <Styled
    styles={
      Object {
        "root": Array [
          Object {
            "selectors": Object {
              "::before": Object {
                "background": undefined,
              },
            },
          },
        ],
      }
    }
  />
  <Stack
    styles={
      Object {
        "root": Object {
          "width": 600,
        },
      }
    }
  >
    <CollapsibleSectionComponent
      isExpandedByDefault={true}
      title="Index(es) to be dropped"
    />
  </Stack>
</Stack>
`;
